libs/utils/tests/logger_plain_test.rs

```diff
@@ -0,0 +1,43 @@
+use std::fs::{File, OpenOptions};
+use std::io::{self, BufRead};
+use tempfile::*;
+use tracing::*;
+use utils::logging::*;
+
+fn read_lines(file: File) -> io::Lines<io::BufReader<File>> {
+    io::BufReader::new(file).lines()
+}
+
+#[test]
+fn test_plain_format_has_message_and_custom_field() {
+    std::env::set_var("RUST_LOG", "warn");
+
+    let tmp_dir = TempDir::new().unwrap();
+    let log_file_path = tmp_dir.path().join("logfile");
+
+    let log_format = LogFormat::Plain;
+    let _log_file = init(&log_file_path, true, log_format).unwrap();
+    assert!(log_file_path.exists());
+
+    let custom_field: &str = "hi";
+    trace!(custom = %custom_field, "test log message");
+    debug!(custom = %custom_field, "test log message");
+    info!(custom = %custom_field, "test log message");
+    warn!(custom = %custom_field, "test log message");
+    error!(custom = %custom_field, "test log message");
+
+    let log_file = OpenOptions::new().read(true).open(&log_file_path).unwrap();
+
+    let lines = read_lines(log_file);
+    for line in lines {
+        let content = line.unwrap();
+        let json = serde_json::from_str::<serde_json::Value>(&content);
+        assert!(!json.is_ok());
+        assert!(content.contains("custom=hi"));
+        assert!(content.contains("test log message"));
+
+        assert!(!content.contains("TRACE"));
+        assert!(!content.contains("DEBUG"));
+        assert!(!content.contains("INFO"));
+    }
+}
```

Comment:
The two test cases should be moved to `libs/utils/src/logging.rs`, either as-is cut-paste or under an additional module:

```
#[cfg(test)]
mod tests {
  #[test]
  fn json_format_has_message_and_custom_field() {
    ...
  }
  
  #[test]
  fn plain_format_has_message_and_custom_field() {
    ...
  }
}
```

The later version doesn't repeat the `test` in the function name as there's the `#[test]` marker already.

The rationale is that each file under `libs/utils/tests/` get a full executable, executing all `#[test]` in the file whereas tests within `src/` get baked into a single executable, linked only once. While for `libs/utils/` there's a much shorter linking time compared to pageserver for example, I'd still tread lightly around making it longer.

The other "benefits" for `libs/utils/tests` are not a factor for these kinds of tests (having access only to the pub API).

Reply:
This was what I tried to do at first and it worked fine until running the tests at the same time. Since we use [init](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/fmt/fn.init.html), the other test just panics. Not sure if we should do something else than initialize a global one.

Reply:
Ah, didn't see that. Yeah, I think I can see a path to not using these different test executables for this, something along the lines of:
1. refactor it so that we have a separate fn to construct a Subscriber
2. use a [tracing::WithSubscriber](https://docs.rs/tracing/0.1.37/tracing/instrument/trait.WithSubscriber.html#method.with_subscriber) to wrap a future with it
3. somehow without bringing in tokio poll the future to completion

Last step alone is way too much legwork for this, so I'd say let's just take the `cargo test` linker hit, until someone gets so frustrated with it to fix it.

Could include a comment on one of the files explaining why this is (different process per executable, executable per file under `tests/`), then refer to it from the other test file.

Reply:
Actually thought, maybe the logger initialization could be macroed away from the test files too: https://github.com/neondatabase/neon/pull/2624/commits/27c5fc5976332bcde39a613daab56b156d46f090

Reply:
Looking good, forgot to resolve this.